route_verification/bgp/src/report/all.rs

```diff
@@ -13,78 +11,63 @@ pub trait ToAnyReport {
 impl ToAnyReport for AllReport {
     fn to_any(self) -> AnyReport {
         match self {
-            Ok(OkT) => None,
-            Ok(SkipT(items)) => Some(SkipF(items)),
-            Ok(MehT(items)) => Some(MehF(items)),
-            Err(items) => Some(BadF(items)),
+            Ok(OkAllReport) => None,
+            Ok(SkipAllReport(items)) => Some(SkipAnyReport(items)),
+            Ok(UnrecAllReport(items)) => Some(UnrecAnyReport(items)),
+            Ok(MehAllReport(items)) => Some(MehAnyReport(items)),
+            Err(items) => Some(BadAnyReport(items)),
         }
     }
 }
 
-pub fn skip_all_report(reason: SkipReason) -> AllReport {
-    let skips = vec![Skip(reason)];
-    Ok(SkipT(skips))
+pub fn skip_all_report(reason: ReportItem) -> AllReport {
+    let skips = vec![reason];
+    Ok(SkipAllReport(skips))
 }
 
 pub const fn empty_skip_all_report() -> AllReport {
-    Ok(SkipT(vec![]))
+    Ok(SkipAllReport(vec![]))
 }
 
-pub fn no_match_all_report(reason: MatchProblem) -> AllReport {
-    let errors = vec![NoMatch(reason)];
+pub fn bad_all_report(reason: ReportItem) -> AllReport {
+    let errors = vec![reason];
     Err(errors)
 }
 
-pub fn bad_rpsl_all_report(reason: RpslError) -> AllReport {
-    let errors = vec![BadRpsl(reason)];
-    Err(errors)
-}
-
-pub fn recursion_all_report(reason: RecurSrc) -> AllReport {
-    let errors = vec![Recursion(reason)];
-    Err(errors)
-}
-
-pub const fn failed_all_report() -> AllReport {
+pub const fn empty_bad_all_report() -> AllReport {
     Err(vec![])
 }
 
-pub enum OkTBad {
-    OkT,
-    SkipT(ReportItems),
-    MehT(ReportItems),
-}
-
-impl OkTBad {
-    pub fn join(self, other: Self) -> Self {
-        match self {
-            OkT => other,
-            SkipT(mut items) => {
-                match other {
-                    OkT => (),
-                    SkipT(i) | MehT(i) => items.extend(i),
-                };
-                SkipT(items)
-            }
-            MehT(mut items) => match other {
-                OkT => MehT(items),
-                SkipT(i) => {
-                    items.extend(i);
-                    SkipT(items)
-                }
-                MehT(i) => {
-                    items.extend(i);
-                    MehT(items)
-                }
-            },
-        }
-    }
+pub enum AllReportCase {
+    OkAllReport,
+    SkipAllReport(ReportItems),
+    UnrecAllReport(ReportItems),
+    MehAllReport(ReportItems),
 }
 
-impl BitAnd for OkTBad {
+impl BitAnd for AllReportCase {
     type Output = Self;
 
-    fn bitand(self, rhs: Self) -> Self::Output {
-        self.join(rhs)
+    /// Merge two `AllReportCase`s based on the rule
+    /// ok → skip → unrec → meh.
+    fn bitand(self, other: Self) -> Self::Output {
+        match (self, other) {
```

Comment:
Maybe it works to `impl PartialOrd for AllReportCase`, and then just sort `[self, other]` to get the best and worse report.